### PR TITLE
New debug option

### PR DESCRIPTION
**Pull request by rodrigosiqueira**

### CODE REVIEW

File: src/debug.sh [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+  # Set default values
+  if [[ -n ${configurations[default_deploy_target]} ]]; then
+    transition_variables=${configurations[default_deploy_target]}
+    options_values['TARGET']=${deploy_target_opt[$transition_variables]}
+    # kw does not support this feature on turn-off vm
+    if [[ "${options_values['TARGET']}" == 1 ]]; then
+      complain 'debug is not supported in VM mode, please change default_deploy_target'
+      exit 95 # EOPNOTSUPP
+    fi
```
Comment (magalilemes) [inline src/debug.sh:420]: This implies that if I set `default_deploy_target` to be `vm`, and then run `kw debug --remote` ... or even `kw debug -h`, I'll get a complaint saying that kw debug doesn't support VM.

**Review by magalilemes — COMMENTED** [review]
An entry briefly describing the command in `kworkflow_help()` in `src/help.sh` is missing.

File: src/remote.sh [hunk 2]
```diff
@@ -52,13 +54,19 @@ function cmd_remotely()
     composed_cmd="ssh -F ${configurations['ssh_configfile']} ${configurations['hostname']}"
   fi
 
+  iterative_output=${iterative_output:-''}
```
Comment (joaoseckler) [inline src/remote.sh:57]: I think this line adds nothing, right? If "$7" is not supplied, then `iterative_output` has the same value as the default, which is `''` (null string).

File: src/remote.sh [hunk 3]
```diff
@@ -40,7 +40,9 @@ function cmd_remotely()
   local port=${4:-${configurations[ssh_port]}}
   local user=${5:-${configurations[ssh_user]}}
   local bash_code="$6"
-  local composed_cmd=""
+  local iterative_output="$7"
```
Comment (joaoseckler) [inline src/remote.sh:43]: I think we should document these options above the function

**Review by joaoseckler — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
@kwy95 could you review the documentation?

File: tests/utils.sh [hunk 4]
```diff
@@ -263,6 +263,25 @@ function create_invalid_file_path()
...
+
+  line=${line:-0}
+
+  equal=$(echo "${expected[@]}" "${result_to_compare[@]}" | tr ' ' '\n' | sort | uniq -u)
+  if [[ -n "$equal" ]]; then
+    assertNull "$line: Arrays are not equal" "$equal"
+    echo "Diff:"
+    echo "-----"
+    echo "$equal"
+    echo "-----"
```
Comment (kwy95) [inline tests/utils.sh:281]: I've slowly started to change change all `echo`s to `printf`s, consider changing them here.
Also, use single quotes where appropriate.

File: src/debug.sh [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+{
+  local list="$1"
+  local type="$2"
+  local ret
+  local msg
+
+  process_list "$list" "$type"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    msg="Something went wrong in the listing process. Please, check if"
```
Comment (kwy95) [inline src/debug.sh:220]: Single quotes

File: src/debug.sh [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+  fi
+}
+
+# The event manipulation has a special syntax for equipping users with a
+# flexible way to gettering different trace options. However, we need a
+# mechanism to translate the user input into a set of paths that can be used
+# for dealing with events. Basically, this function parsers the following
+# syntax:
+#
+# EVENT[:SPECIFIC_EVENT[[filter]][,SPECIFIC_EVENT[[filter]],...]][;EVENT...]
```
Comment (kwy95) [inline src/debug.sh:330]: Use the placeholder syntax mentioned before

File: src/debug.sh [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+
+function parser_debug_options()
+{
+  local options
+  local short_options
+  local long_options
+  local transition_variables
+
+  long_options='remote:,event:,cmd:,local,history,disable,list,follow,help'
+  short_options='f,h,e,c,h,d,l,k'
```
Comment (kwy95) [inline src/debug.sh:412]: `h` appears twice

File: src/debug.sh [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+function debug_help()
+{
+  echo -e "Usage: kw debug [options]:\n" \
+    "\tdebug [--remote [[USER@]REMOTE:[PORT]]] --list [--event TARGET]\n" \
+    "\tdebug [--remote [[USER@]REMOTE:[PORT]]] --history\n" \
+    "\tdebug [--remote [[USER@]REMOTE:[PORT]]] --follow\n" \
+    "\tdebug [--remote [[USER@]REMOTE:[PORT]]] --cmd \"COMMAND\"\n" \
+    "\tdebug [--remote [[USER@]REMOTE:[PORT]]] --event \"SYNTAX\"\n" \
+    "\tYou can combine some of the above options\n"
+}
```
Comment (kwy95) [inline src/debug.sh:520]: Use the new help format with the generated man page and printf

**Review by kwy95 — CHANGES_REQUESTED** [review]
Biggest thing is to separate the documentation on a dedicated file and add it to the man pages,  and then update the help function.
I also think we should discuss a standardization of the error messages, specially on how to display the LINENO variable, as now it is displayed in a myriad of different ways.
And in trying to fulfill #382 consider using printf instead of echo.

**Comment by rodrigosiqueira** [comment]
Hi @kwy95 ,
I address most of your comments. In particular, this update has many improvements in the documentation; please, let me know if the new version looks good for you.

File: tests/utils.sh [hunk 9]
```diff
@@ -263,6 +263,25 @@ function create_invalid_file_path()
...
+function compare_array_values()
+{
+  #shellcheck disable=SC2178
+  local -n expected="$1"
+  local -n result_to_compare="$2"
+  local line="$3"
+
+  line=${line:-0}
+
+  equal=$(echo "${expected[@]}" "${result_to_compare[@]}" | tr ' ' '\n' | sort | uniq -u)
```
Comment (kwy95) [inline tests/utils.sh:275]:
```suggestion
  equal=$(printf '%s\n' "${expected[*]} ${result_to_compare[*]}" | tr ' ' '\n' | sort | uniq -u)
```

File: tests/utils.sh [hunk 10]
```diff
@@ -263,6 +263,25 @@ function create_invalid_file_path()
...
+  if [[ -n "$equal" ]]; then
+    assertNull "$line: Arrays are not equal" "$equal"
+    printf '%s\n' 'Diff:'
+    printf '%s\n' '-----'
+    printf '%s\n' "$equal"
+    printf '%s\n' '-----'
```
Comment (kwy95) [inline tests/utils.sh:281]:
```suggestion
    printf '%s\n' 'Diff:' \
      '-----' \
      "$equal" \
      '-----'
```

File: src/debug.sh [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+  test_mode="${options_values['TEST_MODE']}"
+  target="${options_values['TARGET']}"
+  event="${options_values['EVENT']}"
+  user_cmd="${options_values['CMD']}"
+  keep_history="${options_values['HISTORY']}"
+  disable="${options_values['DISABLE']}"
+  list="${options_values['LIST']}"
+  follow="${options_values['FOLLOW']}"
+
+  # if [[ -n "$event" || (-n "$list" && -z "$event") ]]; then
```
Comment (kwy95) [inline src/debug.sh:62]: Why is this here?

File: src/debug.sh [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+    return
+  fi
+  printf '%s\n' 'kw debug:' \
+    '  debug - kw debug utilities' \
+    '  debug (--remote <remote>:<port> | --local | --vm) - choose target' \
+    '  debug (--list | -l) - List trace options' \
+    '  debug (--history | -h) - Store trace logs in a file' \
+    '  debug (--follow | -f) - Follow traces in real-time' \
+    '  debug (--cmd | -c) \"command\" - Trace log while running a command in the target.' \
+    '  debug (--event | -e) \"<syntax>\" - Trace specific event' \
```
Comment (kwy95) [inline src/debug.sh:525]:
```suggestion
    '  debug (--cmd | -c) "command" - Trace log while running a command in the target.' \
    '  debug (--event | -e) "<syntax>" - Trace specific event' \
```

File: src/debug.sh [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+  if [[ "$1" == --help ]]; then
+    include "$KW_LIB_DIR/help.sh"
+    kworkflow_man 'debug'
+    return
+  fi
+  printf '%s\n' 'kw debug:' \
+    '  debug - kw debug utilities' \
+    '  debug (--remote <remote>:<port> | --local | --vm) - choose target' \
+    '  debug (--list | -l) - List trace options' \
+    '  debug (--history | -h) - Store trace logs in a file' \
```
Comment (kwy95) [inline src/debug.sh:522]: Again, isn't this `-k`?

File: src/debug.sh [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+        ;;
+      --follow | -f)
+        options_values['FOLLOW']=1
+        shift
+        ;;
+      --help | -h)
+        debug_help "$1"
+        exit
+        ;;
+      test_mode)
```
Comment (kwy95) [inline src/debug.sh:496]: This option is not on the list of options passed to `getopt` so its not going to be parsed correctly. Is this even needed? I noticed that `debug_manager()` is not invoked on the test suite.

File: src/debug.sh [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+  local short_options
+  local long_options
+  local transition_variables
+
+  long_options='remote:,event:,cmd:,local,history,disable,list,follow,help'
+  short_options='f,e,c,h,d,l,k'
+
+  options="$(getopt -q --options "$short_options" \
+    --longoptions "$long_options" \
+    -- "$@")"
```
Comment (kwy95) [inline src/debug.sh:416]: Maybe use `kw_parse` instead?

File: src/debug.sh [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+        specific_filter=${specific_filter##*[}
+        specific_event=$(cut -d "[" -f1 <<< "$specific_event")
+      fi
+
+      hash_key="$EVENT_BASE_PATH/$root_event/$specific_event"
+      events_hash["$hash_key"]="$specific_filter"
+    done
+  done
+
+  #if [[ "$option" == "$LIST_OPTION" && -z "$event" ]]; then
```
Comment (kwy95) [inline src/debug.sh:375]: Remove before final version

File: src/debug.sh [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+  fi
+
+  if [[ -n "$event" ]]; then
+    say 'Available events for the requested device:'
+    sed '/^ssh\|^enable\|^filter/d' <<< "$raw_output" | nl
+  fi
+}
+
+# This function is responsible for creating commands that deal with event
+# manipulation, including enabling, disable, and list events.
```
Comment (kwy95) [inline src/debug.sh:258]:
```suggestion
# manipulation, including enabling, disabling, and listing events.
```

File: src/debug.sh [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+#
+# Return:
+# A string with the final command is composed.
+function build_event_command_string()
+{
+  local list="$1"
+  local enable="$2"
+  local enable_events=""
+  local set_filters=""
+  local filter=""
```
Comment (kwy95) [inline src/debug.sh:273]: Single quotes

File: src/debug.sh [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+  local second_time=0
+  local global_trace
+
+  enable=${enable:-"1"}
```
Comment (kwy95) [inline src/debug.sh:277]: Single quotes

**Review by kwy95 — CHANGES_REQUESTED** [review]
Apart from my comments you need to rebase your changes and change the `echo`s with `printf`s. Also review the quote usage, I marked a few of them, but might have missed some as well

**Comment by rodrigosiqueira** [comment]
This PR has been opened for a long time; I decided to merge it since this is working and updated.

**Pull request closed without merging**